src/target/riscv/riscv-013.c

```diff
@@ -4394,182 +4394,280 @@ static int write_memory_bus_v1(struct target *target, target_addr_t address,
-		 * is this read to abstractcs that will cause the dmi_busy_delay
-		 * to be incremented if necessary. */
+	struct riscv_program program;
 
-		uint32_t abstractcs;
-		bool dmi_busy_encountered;
-		result = dm_op(target, &abstractcs, &dmi_busy_encountered,
-				DMI_OP_READ, DM_ABSTRACTCS, 0, false, true);
-		if (result != ERROR_OK)
-			goto error;
-		while (get_field(abstractcs, DM_ABSTRACTCS_BUSY))
-			if (dm_read(target, &abstractcs, DM_ABSTRACTCS) != ERROR_OK)
-				return ERROR_FAIL;
-		info->cmderr = get_field(abstractcs, DM_ABSTRACTCS_CMDERR);
-		if (info->cmderr == CMDERR_NONE && !dmi_busy_encountered) {
-			LOG_TARGET_DEBUG(target, "Successful (partial?) memory write");
-		} else if (info->cmderr == CMDERR_BUSY || dmi_busy_encountered) {
-			if (info->cmderr == CMDERR_BUSY)
-				LOG_TARGET_DEBUG(target, "Memory write resulted in abstract command busy response.");
-			else if (dmi_busy_encountered)
-				LOG_TARGET_DEBUG(target, "Memory write resulted in DMI busy response.");
-			riscv013_clear_abstract_error(target);
-			increase_ac_busy_delay(target);
+	riscv_program_init(&program, target);
+	if (riscv_program_store_mprv(&program, GDB_REGNO_S1, GDB_REGNO_S0, 0, size,
+				mprven) != ERROR_OK)
+		return ERROR_FAIL;
 
-			dm_write(target, DM_ABSTRACTAUTO, 0);
-			result = register_read_direct(target, &cur_addr, GDB_REGNO_S0);
-			if (result != ERROR_OK)
-				goto error;
-			setup_needed = true;
-		} else {
-			LOG_TARGET_ERROR(target, "Error when writing memory, abstractcs=0x%08lx", (long)abstractcs);
-			riscv013_clear_abstract_error(target);
-			result = ERROR_FAIL;
-			goto error;
+	if (riscv_program_addi(&program, GDB_REGNO_S0, GDB_REGNO_S0, size) != ERROR_OK)
+		return ERROR_FAIL;
+
+	if (riscv_program_ebreak(&program) != ERROR_OK)
+		return ERROR_FAIL;
+
+	return riscv_program_write(&program);
+}
+
+static int write_memory_progbuf_inner(struct target *target, target_addr_t start_addr,
+		uint32_t size, uint32_t count, const uint8_t *buffer, bool mprven)
+{
+	if (write_memory_progbuf_fill_progbuf(target, size,
+				mprven) != ERROR_OK)
+		return ERROR_FAIL;
+
+	target_addr_t addr_on_target = start_addr;
+	if (write_memory_progbuf_startup(target, &addr_on_target, buffer, size) != ERROR_OK)
+		return ERROR_FAIL;
+
+	const target_addr_t end_addr = start_addr + (target_addr_t)size * count;
+
+	for (target_addr_t next_addr_on_target = addr_on_target; addr_on_target != end_addr;
+			addr_on_target = next_addr_on_target) {
+		const uint8_t * const curr_buff = buffer + (addr_on_target - start_addr);
+		if (write_memory_progbuf_try_to_write(target, &next_addr_on_target,
+					end_addr, size, curr_buff) != ERROR_OK) {
+			write_memory_progbuf_teardown(target);
+			return ERROR_FAIL;
 		}
+		/* write_memory_progbuf_try_to_write() ensures that at least one item
+		 * gets successfully written even when busy condition is encountered. */
+		assert(next_addr_on_target - addr_on_target > 0);
```

Comment:
Could you write here simply this?

```suggestion
		assert(next_addr_on_target > addr_on_target);
```

Or have you chosen the more complex expression for a specific reason?

Reply:
> Or have you chosen the more complex expression for a specific reason?

Yes. Consider writing on address `(target_addr_t)(-1) - 1` of two bytes. In this case the `next_addr_on_target` will be zero. This will result in the assertion failing.

Reply:
Thanks for explaining. Could you please add a comment about that, e.g. `/* This expression is safe when next_address_on_target overflows */` (or similar).

Reply:
Addressed.